src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php

```diff
@@ -0,0 +1,118 @@
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Attribute\MapQueryString;
+use Symfony\Component\HttpKernel\Attribute\MapRequestPayload;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Serializer\Exception\NotEncodableValueException;
+use Symfony\Component\Serializer\Exception\UnsupportedFormatException;
+use Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Exception\ValidationFailedException;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class RequestPayloadValueResolver implements ValueResolverInterface
+{
+    private const DEFAULT_FORMAT = 'json';
+
+    private const CONTEXT_DENORMALIZE = [
+        AbstractObjectNormalizer::DISABLE_TYPE_ENFORCEMENT => true,
+        DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+    ];
+
+    private const CONTEXT_DESERIALIZE = [
+        DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+    ];
+
+    public function __construct(
+        private readonly ?SerializerInterface $serializer,
+        private readonly ?DenormalizerInterface $denormalizer,
+        private readonly ?ValidatorInterface $validator,
+    ) {
+    }
+
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        $payloadMappers = [
+            MapQueryString::class => [[$this, 'mapQueryString'], Response::HTTP_NOT_FOUND],
+            MapRequestPayload::class => [[$this, 'mapRequestPayload'], Response::HTTP_UNPROCESSABLE_ENTITY],
+        ];
+
+        foreach ($payloadMappers as $mappingAttribute => [$payloadMapper, $validationFailedCode]) {
+            if (!$attributes = $argument->getAttributesOfType($mappingAttribute)) {
+                continue;
+            }
+
+            if (!$type = $argument->getType()) {
+                throw new \LogicException(sprintf('Could not resolve the "$%s" controller argument: argument should be typed.', $argument->getName()));
+            }
+
+            $payload = $payloadMapper($request, $type, $attributes[0]);
+            if (\count($violations = $this->validator?->validate($payload) ?? [])) {
+                throw new ValidationFailedException($payload, $violations, $validationFailedCode);
+            }
```

Comment:
I propose to change the error handling logic this way, and to remove both exception listeners.
This leverages #49944

```diff 
-            $payload = $payloadMapper($request, $type, $attributes[0]);
+            try {
+                $payload = $payloadMapper($request, $type, $attributes[0]);
+            } catch (PartialDenormalizationException $e) {
+                throw new HttpException($validationFailedCode, implode("\n", array_map(fn ($e) => $e->getMessage(), $e->getErrors())), $e);
+            }
+
             if (\count($violations = $this->validator?->validate($payload) ?? [])) {
-                throw new ValidationFailedException($payload, $violations, $validationFailedCode);
+                throw new HttpException($validationFailedCode, $violations, new ValidationFailedException($payload, $violations));
             }
```

Reply:
Let's wait for #49944 then